data/transactions/logic/eval.go

```diff
@@ -4060,3 +4061,62 @@ func opBase64Decode(cx *EvalContext) {
 	}
 	cx.stack[last].Bytes, cx.err = base64Decode(cx.stack[last].Bytes, encoding)
 }
+
+func parseJSON(jsonText []byte) (interface{}, error) {
+	if !json.Valid(jsonText) {
+		return nil, fmt.Errorf("invalid json text %s", jsonText)
+	}
+	var parsed interface{}
+	err := json.Unmarshal(jsonText, &parsed)
+	if err != nil {
+		return nil, err
+	}
+	return parsed, nil
+}
+func opJSONRef(cx *EvalContext) {
+	last := len(cx.stack) - 1
+	fields := strings.Split(string(cx.stack[last].Bytes), ".") // get json keys
+	cx.stack = cx.stack[:last]                                 // pop
+	last = len(cx.stack) - 1
+	resp, err := parseJSON(cx.stack[last].Bytes)
+	if err != nil {
+		cx.err = fmt.Errorf("error while parsing JSON text, %v", err)
+		return
+	}
+	//get value from json
+	for i := 0; i < len(fields); i++ {
+		key := fields[i]
+		switch resp.(type) {
+		case map[string]interface{}:
+			if _, ok := resp.(map[string]interface{})[key]; ok {
+				resp = resp.(map[string]interface{})[key]
+			} else {
```

Comment:
> Are you planning on handling the case of type `[]interface{}` (json array) ? E.g. in [this playground](https://go.dev/play/p/ho3Mxb8M8kI). That would be nice, but it also causes complications because the natural thing to do is to handle array indices numbered parts of the path. In that case, a path such as "key1.4.key2" could be parsed as `["key1", "4", "key2"]` or as `["key1", 4, "key2"]` and one would need to handle both cases (or declare one of them illegal).

Another issue which is probably more fundamental is what if a key itself contains ".". It would be impossible to read its value under the current approach. I have to take some flack here for providing the Python glom example on slack a couple of weeks back.

Reply:
Questions like this make me wonder if we want to support the "." path thing.  If we don't, we probably need to support a return type of "JSONObject" which return a marshalled json string, so json_ref can be used again to dig into it.  But if go this we, it really bugs me that we just can't possibly decode json with "." in the key.  I've definitely seen this (though I think it's gross).

Reply:
If we provide a `split()` opcode, then parsing json with the original approach would be almost as simple as the "." functionality but not cause the headaches...

Reply:
Since AVM has no arrays, split itself is tricky.  Put answers on the stack? With a length? Or maybe a terminating ""?  If the later, what does `pushbytes ".."; split;` do?

Reply:
I think completeness is more important than convenience for this opcode, so I'm included to say the key should **NOT** be interpreted as a path because of the "." issue.

Sure, in most cases fields won't have a "." in them, and perhaps we could provide another opcode to "dig" into a JSON object with a path for convenience (though I don't think it's worth it at this time), but since this opcode version has shortcomings, it should not be our only JSON parsing opcode.

Reply:
I think I feel the same way, but it does mean we need to add another possible extraction, an entire JSONObject, which would be serialized. And we'll have to define that serialization - it needs to have a canonical form beyond "JSON" - we'll have to specify whitespace, key order, escape usage (I think there are multiple ways to escape the same json string).

Unless we can determine that in the first use case, there is no need to dig into an object in the main object.  Sadly, I think that _is_ a requirement, but we can check and skip this for now if it isn't.

Reply:
We'll have to check for edge cases, but Go's `json.RawMessage` seems like a good way to return a subobject "Exactly as it appears in the original json".

Here's an example I made: https://go.dev/play/p/_IXxH7zx1XD

Reply:
Well that looks great!

Reply:
the plan is to push rawmessage to stack and call json_ref again to dig the value, correct?

Reply:
@shiqizng that's correct. `json_ref` should let you extract one level at a time, and we will need `JSONObject` to extract an embedded object.

@jannotti is there a requirement to support JSON array access as well?